doc/dev/design/NAT-address-discovery.rst

```diff
@@ -0,0 +1,163 @@
+-  Time to rollout: changing the border routers should be much easier and faster than getting NAT vendors to implement
+   SCION compatibility or to get rid of NATs completely in home networks.
+   A short time until rollout seems important because people are already running into this problem.
+
+Alternatives:
+
+-  The SRC address/port is updated by the border router to reflect what the border router sees as source address.
+   Problem:
+
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers,
+      e.g. for `SPAO <https://docs.scion.org/en/latest/protocols/authenticator-option.html>`_.
+
+-  The SRC address/port is updated by the NAT. This is similar to having the border router update SRC IP/port.
+
+   - Complicates cryptographically protecting the header if the header must be modifiable by NATs.
+   - Requires cooperation of NAT suppliers to include and roll-out SCION support to all their devices.
+
+-  Use separate STUN servers. This is a possibility, but adds setup complexity and may not work in all cases. Every
+   subnet of an AS that has a border router must also have a STUN server. Moreover, if the STUN server uses a different
+   IP (or port) than the border router, then the NAT may decide to use a different port mapping when connecting to it,
+   i.e. the STUN server may not see the same IP/port tuple on the NAT that the border router sees. Disadvantages:
+
+   -  This approach may be be problematic with sensitive NATs.
+   -  We need to somehow standardize the STUN IP/port and/or communicate it to endhosts, e.g. via the topology.json file
+      or the bootstrapping service.
+
+-  Remove all NATs and use IPv6 instead. This is technically possible but unlikely to happen anytime soon, especially
+   because scarcity of IPv4 addresses is not the only reason why NATs are deployed.
+
+Compatibility
+=============
+[A discussion of breaking changes and how this change can be deployed.]
+
+Breaking changes
+----------------
+
+This change should not break anything.
+
+However, there are some points that may need more discussion:
+
+- Return paths: the proposal requires packets that come from a server to arrive through the same BR that was used for an
+  outgoing request. This currently seems to be the default behavior of servers, but it is not a behavior required
+  by the current standard.
+
+- Dispatcherless port range: What if the the port mapping at the NAT doesn't result in a L4 port that is within the
+  `"dispatched_ports" range <https://docs.scion.org/en/latest/dev/design/router-port-dispatch.html>`_ ?
+  The last-hop BR would then choose the fixed end-host data port (default 30041) as the underlay UDP/IP destination port
+  and NAT mapping would fail. Contrary to point (1.) this problem would at least be detectable by the endhost before
+  based on the proposed NAT IP/port discovery mechanism.
+
```

Comment:
One additional point to check: all client libraries (snet/pan, jpan) have to accept incoming packets where the underlay UDP port does not match with the L4 port, i.e., the underlay port could be the local receiver port whereas the L4 port will be the NATed port.